File: packages/Axiom.jl/zig/src/pool.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
//! Pooling Operations
//!
//! Max pooling, average pooling, and global pooling.

const std = @import("std");
const math = std.math;

/// 2D Max Pooling
pub fn maxpool2d(
    input: [*]const f32,
    output: [*]f32,
    batch: usize,
    h_in: usize,
    w_in: usize,
    channels: usize,
    kh: usize,
    kw: usize,
    stride_h: usize,
    stride_w: usize,
) void {
    const h_out = (h_in - kh) / stride_h + 1;
    const w_out = (w_in - kw) / stride_w + 1;
    const input_hw = h_in * w_in;
    const output_hw = h_out * w_out;

    var n: usize = 0;
    while (n < batch) : (n += 1) {
        const input_batch = input + n * input_hw * channels;
        const output_batch = output + n * output_hw * channels;

        var c: usize = 0;
        while (c < channels) : (c += 1) {
            var oh: usize = 0;
            while (oh < h_out) : (oh += 1) {
                var ow: usize = 0;
                while (ow < w_out) : (ow += 1) {
                    var max_val: f32 = -math.floatMax(f32);

                    var ki: usize = 0;
                    while (ki < kh) : (ki += 1) {
                        var kj: usize = 0;
                        while (kj < kw) : (kj += 1) {
                            const ih = oh * stride_h + ki;
                            const iw = ow * stride_w + kj;
                            const in_idx = ih * w_in * channels + iw * channels + c;
                            const val = input_batch[in_idx];
                            if (val > max_val) max_val = val;
                        }
                    }

                    const out_idx = oh * w_out * channels + ow * channels + c;
                    output_batch[out_idx] = max_val;
                }
            }
        }
    }
}

/// 2D Average Pooling
pub fn avgpool2d(
    input: [*]const f32,
    output: [*]f32,
    batch: usize,
    h_in: usize,
    w_in: usize,
    channels: usize,
    kh: usize,
    kw: usize,
    stride_h: usize,
    stride_w: usize,
) void {
    const h_out = (h_in - kh) / stride_h + 1;
    const w_out = (w_in - kw) / stride_w + 1;
    const input_hw = h_in * w_in;
    const output_hw = h_out * w_out;
    const kernel_size: f32 = @floatFromInt(kh * kw);

    var n: usize = 0;
    while (n < batch) : (n += 1) {
        const input_batch = input + n * input_hw * channels;
        const output_batch = output + n * output_hw * channels;

        var c: usize = 0;
        while (c < channels) : (c += 1) {
            var oh: usize = 0;
            while (oh < h_out) : (oh += 1) {
                var ow: usize = 0;
                while (ow < w_out) : (ow += 1) {
                    var sum: f32 = 0;

                    var ki: usize = 0;
                    while (ki < kh) : (ki += 1) {
                        var kj: usize = 0;
                        while (kj < kw) : (kj += 1) {
                            const ih = oh * stride_h + ki;
                            const iw = ow * stride_w + kj;
                            const in_idx = ih * w_in * channels + iw * channels + c;
                            sum += input_batch[in_idx];
                        }
                    }

                    const out_idx = oh * w_out * channels + ow * channels + c;
                    output_batch[out_idx] = sum / kernel_size;
                }
            }
        }
    }
}

/// Global Average Pooling
/// Reduces (N, H, W, C) to (N, C)
pub fn global_avgpool2d(
    input: [*]const f32,
    output: [*]f32,
    batch: usize,
    h: usize,
    w: usize,
    channels: usize,
) void {
    const spatial_size = h * w;
    const spatial_size_f: f32 = @floatFromInt(spatial_size);

    var n: usize = 0;
    while (n < batch) : (n += 1) {
        const input_batch = input + n * spatial_size * channels;

        var c: usize = 0;
        while (c < channels) : (c += 1) {
            var sum: f32 = 0;

            var s: usize = 0;
            while (s < spatial_size) : (s += 1) {
                sum += input_batch[s * channels + c];
            }

            output[n * channels + c] = sum / spatial_size_f;
        }
    }
}

/// Global Max Pooling
/// Reduces (N, H, W, C) to (N, C)
pub fn global_maxpool2d(
    input: [*]const f32,
    output: [*]f32,
    batch: usize,
    h: usize,
    w: usize,
    channels: usize,
) void {
    const spatial_size = h * w;

    var n: usize = 0;
    while (n < batch) : (n += 1) {
        const input_batch = input + n * spatial_size * channels;

        var c: usize = 0;
        while (c < channels) : (c += 1) {
            var max_val: f32 = -math.floatMax(f32);

            var s: usize = 0;
            while (s < spatial_size) : (s += 1) {
                const val = input_batch[s * channels + c];
                if (val > max_val) max_val = val;
            }

            output[n * channels + c] = max_val;
        }
    }
}

/// Adaptive Average Pooling
/// Produces fixed output size regardless of input
pub fn adaptive_avgpool2d(
    input: [*]const f32,
    output: [*]f32,
    batch: usize,
    h_in: usize,
    w_in: usize,
    channels: usize,
    h_out: usize,
    w_out: usize,
) void {
    const input_hw = h_in * w_in;
    const output_hw = h_out * w_out;

    var n: usize = 0;
    while (n < batch) : (n += 1) {
        const input_batch = input + n * input_hw * channels;
        const output_batch = output + n * output_hw * channels;

        var c: usize = 0;
        while (c < channels) : (c += 1) {
            var oh: usize = 0;
            while (oh < h_out) : (oh += 1) {
                var ow: usize = 0;
                while (ow < w_out) : (ow += 1) {
                    // Compute window bounds
                    const h_start = oh * h_in / h_out;
                    const h_end = (oh + 1) * h_in / h_out;
                    const w_start = ow * w_in / w_out;
                    const w_end = (ow + 1) * w_in / w_out;

                    var sum: f32 = 0;
                    var count: usize = 0;

                    var ih = h_start;
                    while (ih < h_end) : (ih += 1) {
                        var iw = w_start;
                        while (iw < w_end) : (iw += 1) {
                            sum += input_batch[ih * w_in * channels + iw * channels + c];
                            count += 1;
                        }
                    }

                    const count_f: f32 = @floatFromInt(count);
                    output_batch[oh * w_out * channels + ow * channels + c] = sum / count_f;
                }
            }
        }
    }
}

File: packages/Axiom.jl/rust/src/ops/pool.rs
//! Pooling operations

use ndarray::{Array4, ArrayView4};
use rayon::prelude::*;

/// 2D Max Pooling
///
/// Input shape: (N, H, W, C)
/// Output shape: (N, H_out, W_out, C)
pub fn maxpool2d(
    input: ArrayView4<f32>,
    kernel_size: (usize, usize),
    stride: (usize, usize),
    padding: (usize, usize),
) -> Array4<f32> {
    let (n, h_in, w_in, c) = input.dim();
    let (kh, kw) = kernel_size;
    let (sh, sw) = stride;
    let (ph, pw) = padding;

    let h_out = (h_in + 2 * ph - kh) / sh + 1;
    let w_out = (w_in + 2 * pw - kw) / sw + 1;

    let mut output = Array4::from_elem((n, h_out, w_out, c), f32::NEG_INFINITY);

    output
        .axis_iter_mut(ndarray::Axis(0))
        .into_par_iter()
        .enumerate()
        .for_each(|(batch_idx, mut batch_output)| {
            for channel in 0..c {
                for i in 0..h_out {
                    for j in 0..w_out {
                        let h_start = i * sh;
                        let w_start = j * sw;

                        let mut max_val = f32::NEG_INFINITY;

                        for ki in 0..kh {
                            for kj in 0..kw {
                                let h_idx = h_start + ki;
                                let w_idx = w_start + kj;

                                if h_idx >= ph
                                    && h_idx < h_in + ph
                                    && w_idx >= pw
                                    && w_idx < w_in + pw
                                {
                                    let val = input[[batch_idx, h_idx - ph, w_idx - pw, channel]];
                                    if val > max_val {
                                        max_val = val;
                                    }
                                }
                            }
                        }

                        batch_output[[i, j, channel]] = max_val;
                    }
                }
            }
        });

    output
}

/// 2D Max Pooling with indices (for unpooling)
pub fn maxpool2d_with_indices(
    input: ArrayView4<f32>,
    kernel_size: (usize, usize),
    stride: (usize, usize),
    padding: (usize, usize),
) -> (Array4<f32>, Array4<usize>) {
    let (n, h_in, w_in, c) = input.dim();
    let (kh, kw) = kernel_size;
    let (sh, sw) = stride;
    let (ph, pw) = padding;

    let h_out = (h_in + 2 * ph - kh) / sh + 1;
    let w_out = (w_in + 2 * pw - kw) / sw + 1;

    let mut output = Array4::from_elem((n, h_out, w_out, c), f32::NEG_INFINITY);
    let mut indices = Array4::zeros((n, h_out, w_out, c));

    for batch_idx in 0..n {
        for channel in 0..c {
            for i in 0..h_out {
                for j in 0..w_out {
                    let h_start = i * sh;
                    let w_start = j * sw;

                    let mut max_val = f32::NEG_INFINITY;
                    let mut max_idx = 0usize;

                    for ki in 0..kh {
                        for kj in 0..kw {
                            let h_idx = h_start + ki;
                            let w_idx = w_start + kj;

                            if h_idx >= ph && h_idx < h_in + ph && w_idx >= pw && w_idx < w_in + pw
                            {
                                let linear_idx = (h_idx - ph) * w_in + (w_idx - pw);
                                let val = input[[batch_idx, h_idx - ph, w_idx - pw, channel]];

                                if val > max_val {
                                    max_val = val;
                                    max_idx = linear_idx;
                                }
                            }
                        }
                    }

                    output[[batch_idx, i, j, channel]] = max_val;
                    indices[[batch_idx, i, j, channel]] = max_idx;
                }
            }
        }
    }

    (output, indices)
}

/// 2D Average Pooling
pub fn avgpool2d(
    input: ArrayView4<f32>,
    kernel_size: (usize, usize),
    stride: (usize, usize),
    padding: (usize, usize),
    count_include_pad: bool,
) -> Array4<f32> {
    let (n, h_in, w_in, c) = input.dim();
    let (kh, kw) = kernel_size;
    let (sh, sw) = stride;
    let (ph, pw) = padding;

    let h_out = (h_in + 2 * ph - kh) / sh + 1;
    let w_out = (w_in + 2 * pw - kw) / sw + 1;

    let mut output = Array4::zeros((n, h_out, w_out, c));

    output
        .axis_iter_mut(ndarray::Axis(0))
        .into_par_iter()
        .enumerate()
        .for_each(|(batch_idx, mut batch_output)| {
            for channel in 0..c {
                for i in 0..h_out {
                    for j in 0..w_out {
                        let h_start = i * sh;
                        let w_start = j * sw;

                        let mut sum = 0.0f32;
                        let mut count = 0usize;

                        for ki in 0..kh {
                            for kj in 0..kw {
                                let h_idx = h_start + ki;
                                let w_idx = w_start + kj;

                                if h_idx >= ph
                                    && h_idx < h_in + ph
                                    && w_idx >= pw
                                    && w_idx < w_in + pw
                                {
                                    sum += input[[batch_idx, h_idx - ph, w_idx - pw, channel]];
                                    count += 1;
                                } else if count_include_pad {
                                    count += 1;
                                }
                            }
                        }

                        let divisor = if count_include_pad { kh * kw } else { count };
                        batch_output[[i, j, channel]] = sum / divisor as f32;
                    }
                }
            }
        });

    output
}

/// Global Average Pooling
pub fn global_avgpool2d(input: ArrayView4<f32>) -> ndarray::Array2<f32> {
    let (n, h, w, c) = input.dim();
    let mut output = ndarray::Array2::zeros((n, c));

    for batch_idx in 0..n {
        for channel in 0..c {
            let mut sum = 0.0f32;
            for i in 0..h {
                for j in 0..w {
                    sum += input[[batch_idx, i, j, channel]];
                }
            }
            output[[batch_idx, channel]] = sum / (h * w) as f32;
        }
    }

    output
}

/// Global Max Pooling
pub fn global_maxpool2d(input: ArrayView4<f32>) -> ndarray::Array2<f32> {
    let (n, h, w, c) = input.dim();
    let mut output = ndarray::Array2::from_elem((n, c), f32::NEG_INFINITY);

    for batch_idx in 0..n {
        for channel in 0..c {
            for i in 0..h {
                for j in 0..w {
                    let val = input[[batch_idx, i, j, channel]];
                    if val > output[[batch_idx, channel]] {
                        output[[batch_idx, channel]] = val;
                    }
                }
            }
        }
    }

    output
}

/// Adaptive Average Pooling
pub fn adaptive_avgpool2d(input: ArrayView4<f32>, output_size: (usize, usize)) -> Array4<f32> {
    let (n, h_in, w_in, c) = input.dim();
    let (h_out, w_out) = output_size;

    let mut output = Array4::zeros((n, h_out, w_out, c));

    for batch_idx in 0..n {
        for channel in 0..c {
            for i in 0..h_out {
                for j in 0..w_out {
                    // Compute window
                    let h_start = (i * h_in) / h_out;
                    let h_end = ((i + 1) * h_in).div_ceil(h_out);
                    let w_start = (j * w_in) / w_out;
                    let w_end = ((j + 1) * w_in).div_ceil(w_out);

                    let mut sum = 0.0f32;
                    let mut count = 0usize;

                    for hi in h_start..h_end {
                        for wi in w_start..w_end {
                            sum += input[[batch_idx, hi, wi, channel]];
                            count += 1;
                        }
                    }

                    output[[batch_idx, i, j, channel]] = sum / count as f32;
                }
            }
        }
    }

    output
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_maxpool2d() {
        let input =
            Array4::from_shape_vec((1, 4, 4, 1), (0..16).map(|x| x as f32).collect()).unwrap();

        let output = maxpool2d(input.view(), (2, 2), (2, 2), (0, 0));

        assert_eq!(output.dim(), (1, 2, 2, 1));
        assert_eq!(output[[0, 0, 0, 0]], 5.0); // max of [0,1,4,5]
        assert_eq!(output[[0, 1, 1, 0]], 15.0); // max of [10,11,14,15]
    }

    #[test]
    fn test_global_avgpool() {
        let input = Array4::from_elem((2, 4, 4, 3), 1.0f32);
        let output = global_avgpool2d(input.view());

        assert_eq!(output.dim(), (2, 3));
        assert_eq!(output[[0, 0]], 1.0);
    }
}
